### PR TITLE
Feat/add stake wars section

### PR DESCRIPTION
**Pull request by dnoishi**
## **Type**
enhancement, documentation

___

## **Description**
- Introduced components and infrastructure for the Stake Wars section.
- Added `StakeWars`, `EndGame`, `NominatorList`, and `OperatorsList` components.
- Defined GraphQL queries for fetching data related to Stake Wars.
- Updated `ChainProvider` to handle multiple GraphQL endpoints.
- Added constants and routes specific to Stake Wars.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>11 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>index.tsx</strong><dd><code>Add Stake Wars Leaderboard Component</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

explorer/src/components/StakeWars/index.tsx
<li>Added <code>StakeWars</code> component to display the leaderboard for Stake Wars.<br> <li> Utilizes <code>PageTabs</code> for navigating between different phases and <code>EndGame</code>.<br> <li> Includes queries to fetch current block number and determine phase <br>state.<br>

</details>
    

  </td>
  <td><a href="https://github.com/subspace/astral/pull/565/files#diff-b7bcc7224f24915a0c30b4f69d68f49cb67c3dcfb3c505c07b6a4d0b79d9eeba">+137/-0</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>EndGame.tsx</strong><dd><code>Implement End Game Component for Stake Wars</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

explorer/src/components/StakeWars/EndGame.tsx
<li>Introduced <code>EndGame</code> component to display winners of the Stake Wars.<br> <li> Displays the operator and nominator with the highest rewards.<br> <li> Implements fetching and calculation of rewards for display.<br>

</details>
    

  </td>
  <td><a href="https://github.com/subspace/astral/pull/565/files#diff-8a086b3883153a21496924a792b05bb6bec3a97ac4f0f726ff93ee643b910301">+162/-0</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>NominatorList.tsx</strong><dd><code>Add Nominator List Component for Stake Wars</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

explorer/src/components/StakeWars/NominatorList.tsx
<li>Created <code>NominatorList</code> component to list nominators in Stake Wars.<br> <li> Supports pagination, searching, and sorting of nominator data.<br> <li> Integrates with GraphQL to fetch nominator information.<br>

</details>
    

  </td>
  <td><a href="https://github.com/subspace/astral/pull/565/files#diff-b37b0942138d21a30863ee7f88fc2129f3e8df64a2817192598c9e7075d2ce44">+192/-0</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>OperatorsList.tsx</strong><dd><code>Add Operators List Component for Stake Wars</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

explorer/src/components/StakeWars/OperatorsList.tsx
<li>Added <code>OperatorsList</code> component for listing operators in Stake Wars.<br> <li> Features include pagination, search, and sorting capabilities.<br> <li> Connects with GraphQL for data retrieval.<br>

</details>
    

  </td>
  <td><a href="https://github.com/subspace/astral/pull/565/files#diff-486f7063212463bcdb3375b9db7450dedd2b997d8a558bfc40192f15a319ed99">+224/-0</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>query.ts</strong><dd><code>Define GraphQL Queries for Stake Wars Data</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

explorer/src/components/StakeWars/query.ts
<li>Defined GraphQL queries for fetching Stake Wars data.<br> <li> Includes queries for current block number, operators, and nominators.<br>

</details>
    

  </td>
  <td><a href="https://github.com/subspace/astral/pull/565/files#diff-e5ad8acf572f4d04102b6c044bd22feafc43d91b91aaf7518e5b1f6708281ca7">+130/-0</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>general.ts</strong><dd><code>Add Constants for Stake Wars</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

explorer/src/constants/general.ts
- Added constants for Stake Wars phases and page size.

</details>
    

  </td>
  <td><a href="https://github.com/subspace/astral/pull/565/files#diff-3003cef0777f86baed95989b26937ac637617cce3073d73e20d874869d888fb3">+17/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>ChainProvider.tsx</strong><dd><code>Update ChainProvider for Multiple Endpoints</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

explorer/src/providers/ChainProvider.tsx
<li>Updated <code>ChainProvider</code> to support multiple endpoints based on context.<br> <li> Allows fetching data from different GraphQL endpoints for Stake Wars.<br>

</details>
    

  </td>
  <td><a href="https://github.com/subspace/astral/pull/565/files#diff-a55a9b593552c3724c0ea05e726001ec210930c00c73a6549093b80abe4f7ced">+26/-2</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>StakeHeader.tsx</strong><dd><code>Implement StakeHeader Component for Navigation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

explorer/src/components/layout/StakeHeader.tsx
<li>Introduced <code>StakeHeader</code> component for the Stake Wars section.<br> <li> Provides navigation and theme toggling specific to Stake Wars.<br>

</details>
    

  </td>
  <td><a href="https://github.com/subspace/astral/pull/565/files#diff-7f446cc0b7ef7044f629f33c394cf56a0d2c01d1cd6d4610ec570a252283834c">+81/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>routes.ts</strong><dd><code>Add Route Constant for Stake Wars</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

explorer/src/constants/routes.ts
- Added route constant for Stake Wars section.

</details>
    

  </td>
  <td><a href="https://github.com/subspace/astral/pull/565/files#diff-1db40c1c0135e9c55c149a79c123aafb3fc2408f35965369fe6ff07598b37d91">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Badge.tsx</strong><dd><code>Add Badge Component for Phase State Display</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

explorer/src/components/common/Badge.tsx
- Added `Badge` component for displaying phase states in Stake Wars.

</details>
    

  </td>
  <td><a href="https://github.com/subspace/astral/pull/565/files#diff-994872d3a4c8512d1e8058260fe647293f1e6768c2843d9fe74946a1d85953b0">+30/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>NotStarted.tsx</strong><dd><code>Implement NotStarted Component for Stake Wars</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

explorer/src/components/layout/NotStarted.tsx
<li>Created <code>NotStarted</code> component for phases of Stake Wars that haven't <br>started.<br>

</details>
    

  </td>
  <td><a href="https://github.com/subspace/astral/pull/565/files#diff-9fed2c9c4c8ef14bfac21ffd3dad0889656a0f716bf2154d2a6cac993a595e39">+21/-0</a>&nbsp; &nbsp; </td>
</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: explorer/src/components/StakeWars/index.tsx [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+import { useErrorHandler } from 'react-error-boundary'
+import Badge from '../common/Badge'
+import { EndGame } from './EndGame'
+import { NominatorList } from './NominatorList'
+import { OperatorsList } from './OperatorsList'
+import { phaseState } from './helpers/phaseState'
+import { GET_CURRENT_BLOCK_NUMBER } from './query'
+
+const StakeWars = () => {
+  const { data, loading, error } = useQuery(GET_CURRENT_BLOCK_NUMBER, {
```
Comment (marc-aurele-besner) [inline explorer/src/components/StakeWars/index.tsx:18]:
```suggestion
  const { data, loading, error } = useQuery<GetCurrentBlockNumberQuery>(GET_CURRENT_BLOCK_NUMBER, {
```

File: explorer/src/components/StakeWars/OperatorsList.tsx [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+          ? (pagination.pageIndex * pagination.pageSize).toString()
+          : undefined,
+      orderBy: sorting.map((s) => `${s.id}_${s.desc ? 'DESC' : 'ASC'}`).join(',') || 'id_ASC',
+      blockNumber_gte: STAKE_WARS_PHASES.phase2.start,
+      blockNumber_lte: STAKE_WARS_PHASES.phase2.end,
+    }),
+    [pagination],
+  )
+
+  const { data, error, loading } = useQuery(GET_ALL_OPERATORS, {
```
Comment (marc-aurele-besner) [inline explorer/src/components/StakeWars/OperatorsList.tsx:126]:
```suggestion
  const { data, error, loading } = useQuery<GetAllOperatorsQuery>(GET_ALL_OPERATORS, {
```

File: explorer/src/components/StakeWars/OperatorsList.tsx [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+import { SortingState } from '@tanstack/react-table'
+import { OperatorsListCard } from 'components/StakeWars/OperatorListCard'
+import { GET_ALL_OPERATORS } from 'components/StakeWars/query'
+import { DebouncedInput } from 'components/common/DebouncedInput'
+import { NewTable } from 'components/common/NewTable'
+import { Spinner } from 'components/common/Spinner'
+import { NotFound } from 'components/layout/NotFound'
+import { PAGE_SIZE, STAKE_WARS_PHASES } from 'constants/'
+import { INTERNAL_ROUTES } from 'constants/routes'
+import type { OperatorsConnectionQuery } from 'gql/graphql'
```
Comment (marc-aurele-besner) [inline explorer/src/components/StakeWars/OperatorsList.tsx:16]:
```suggestion
import type { OperatorsConnectionQuery, GetAllOperatorsQuery } from 'gql/graphql'
```

File: explorer/src/components/StakeWars/NominatorList.tsx [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+          : undefined,
+      orderBy: sorting.map((s) => `${s.id}_${s.desc ? 'DESC' : 'ASC'}`).join(',') || 'id_ASC',
+      where: searchOperator ? { id_eq: searchOperator } : {},
+      blockNumber_gte: STAKE_WARS_PHASES.phase3.start,
+      blockNumber_lte: STAKE_WARS_PHASES.phase3.end,
+    }),
+    [sorting, pagination, searchOperator],
+  )
+
+  const { data, error, loading } = useQuery(GET_ALL_NOMINATORS, {
```
Comment (marc-aurele-besner) [inline explorer/src/components/StakeWars/NominatorList.tsx:90]:
```suggestion
  const { data, error, loading } = useQuery<GetAllNominatorsQuery>(GET_ALL_NOMINATORS, {
```

File: explorer/src/components/StakeWars/NominatorList.tsx [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+import { shortString } from '@/utils/string'
+import { useApolloClient, useQuery } from '@apollo/client'
+import { SortingState } from '@tanstack/react-table'
+import { GET_ALL_NOMINATORS } from 'components/StakeWars/query'
+import { DebouncedInput } from 'components/common/DebouncedInput'
+import { NewTable } from 'components/common/NewTable'
+import { Spinner } from 'components/common/Spinner'
+import { NotFound } from 'components/layout/NotFound'
+import { STAKE_WARS_PAGE_SIZE, STAKE_WARS_PHASES } from 'constants/'
+import type { NominatorsConnectionQuery } from 'gql/graphql'
```
Comment (marc-aurele-besner) [inline explorer/src/components/StakeWars/NominatorList.tsx:14]:
```suggestion
import type { NominatorsConnectionQuery, GetAllNominatorsQuery } from 'gql/graphql'
```

File: explorer/src/components/StakeWars/EndGame.tsx [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+      after: undefined,
+      orderBy: sorting.map((s) => `${s.id}_${s.desc ? 'DESC' : 'ASC'}`).join(',') || 'id_ASC',
+      where: {},
+      blockNumber_gte: STAKE_WARS_PHASES.phase2.start,
+      blockNumber_lte: STAKE_WARS_PHASES.phase2.end,
+    }),
+    [sorting, pagination],
+  )
+
+  const { data, error, loading } = useQuery(GET_ALL_NOMINATORS, {
```
Comment (marc-aurele-besner) [inline explorer/src/components/StakeWars/EndGame.tsx:41]:
```suggestion
  const { data, error, loading } = useQuery<GetAllNominatorsQuery>(GET_ALL_NOMINATORS, {
```

File: explorer/src/components/StakeWars/EndGame.tsx [hunk 7]
```diff
@@ -0,0 +1,162 @@
+/* eslint-disable camelcase */
+'use client'
+
+import { bigNumberToNumber } from '@/utils/number'
+import { capitalizeFirstLetter, shortString } from '@/utils/string'
+import { useQuery } from '@apollo/client'
+import { SortingState } from '@tanstack/react-table'
```
Comment (marc-aurele-besner) [inline explorer/src/components/StakeWars/EndGame.tsx:7]:
```suggestion
import { SortingState } from '@tanstack/react-table'
import type { GetAllNominatorsQuery } from 'gql/graphql'
```

**Review by marc-aurele-besner — CHANGES_REQUESTED** [review]
Awesome, looking great!
I suggested a few changes, primarily to improve type safety.

**Review by marc-aurele-besner — APPROVED** [review]
LGTM 🚀

**Review by marc-aurele-besner — DISMISSED** [review]

**Review by marc-aurele-besner — COMMENTED** [review]
Screenshot 2024-04-17 at 5 18 35 PM

The count does not match the amount of rows in the first section (stake)

image

The account is shown in the id column and in the account column, instead of a increment or id. for the second section (nominators)

**Comment by marc-aurele-besner** [comment]
I'm guessing we want this pr to target staging first :)

**Comment by jfrank-summit** [comment]
We need to be able to see all of the rewards, not just the top accounts. It also needs to be exportable so that we can make a proper accounting of mainnet tokens earned.

**Review by marc-aurele-besner — APPROVED** [review]

**Comment by jfrank-summit** [comment]
> We need to be able to see all of the rewards, not just the top accounts. It also needs to be exportable so that we can make a proper accounting of mainnet tokens earned.

@marc-aurele-besner was this addressed? The preview doesn't work anymore.

**Comment by marc-aurele-besner** [comment]
> > We need to be able to see all of the rewards, not just the top accounts. It also needs to be exportable so that we can make a proper accounting of mainnet tokens earned.
> 
> @marc-aurele-besner was this addressed? The preview doesn't work anymore.

My last commit address it, but in a sloppy way.
It's now displaying the page navigation and download button, but if we end up with more rows than the query limit size (150) then the amount of pages will be limited to the amount of row the query return.

I created a separate issue, [to improve the logic of the Table components](https://github.com/subspace/astral/issues/582)

**Review by jfrank-summit — APPROVED** [review]

**Pull request merged**